Mathlib/GroupTheory/SpecificGroups/Cyclic.lean

```diff
@@ -787,3 +791,105 @@ noncomputable def mulEquivOfPrimeCardEq {p : ℕ} [Fintype G] [Fintype H] [Group
+  map_one' := orderOf_dvd_iff_zpow_eq_one.mp <|
+                (Int.natCast_dvd_natCast.mpr hg').trans <| orderOf_dvd_iff_zpow_eq_one.mpr <|
+                Classical.choose_spec <| mem_zpowers_iff.mp <| hg 1
+  map_mul' x y := by
+    simp only [← zpow_add, zpow_eq_zpow_iff_modEq]
+    apply Int.ModEq.of_dvd (Int.natCast_dvd_natCast.mpr hg')
+    rw [← zpow_eq_zpow_iff_modEq, zpow_add]
+    simp only [fun x ↦ Classical.choose_spec <| mem_zpowers_iff.mp <| hg x]
+
+@[to_additive]
+lemma monoidHom_of_forall_mem_zpowers_apply_gen :
+    monoidHom_of_forall_mem_zpowers hg hg' g = g' := by
+  simp only [monoidHom_of_forall_mem_zpowers, MonoidHom.coe_mk, OneHom.coe_mk]
+  nth_rw 2 [← zpow_one g']
+  rw [zpow_eq_zpow_iff_modEq]
+  apply Int.ModEq.of_dvd (Int.natCast_dvd_natCast.mpr hg')
+  rw [← zpow_eq_zpow_iff_modEq, zpow_one]
+  exact Classical.choose_spec <| mem_zpowers_iff.mp <| hg g
+
+end monoidHom
+
+/-- Two group homomorphisms `G →* G'` are equal if and only if they agree on a generator of `G`. -/
+@[to_additive
+   "Two homomorphisms `G →+ G'` of additive groups are equal if and only if they agree
+   on a generator of `G`."]
+lemma MonoidHom.eq_iff_eq_on_generator (f₁ f₂ : G →* G') : f₁ = f₂ ↔ f₁ g = f₂ g := by
+  rw [DFunLike.ext_iff]
+  refine ⟨fun H ↦ H g, fun H x ↦ ?_⟩
+  obtain ⟨n, hn⟩ := mem_zpowers_iff.mp <| hg x
+  rw [← hn, map_zpow, map_zpow, H]
+
+/-- Two group isomorphisms `G ≃* G'` are equal if and only if they agree on a generator of `G`. -/
+@[to_additive
+   "Two isomorphisms `G ≃+ G'` of additive groups are equal if and only if they agree
+   on a generator of `G`."]
+lemma MulEquiv.eq_iff_eq_on_generator (f₁ f₂ : G ≃* G') : f₁ = f₂ ↔ f₁ g = f₂ g :=
+  (Function.Injective.eq_iff toMonoidHom_injective).symm.trans <|
+    MonoidHom.eq_iff_eq_on_generator hg ..
+
+section mulEquiv
+
+variable (hg' : ∀ x, x ∈ zpowers g') (h : orderOf g = orderOf g')
+
+/-- Given two groups that are generated by elements `g` and `g'` of the same order,
+we obtain an isomorphism sending `g` to `g'`. -/
+@[to_additive
+   "Given two additive groups that are generated by elements `g` and `g'` of the same order,
+   we obtain an isomorphism sending `g` to `g'`."]
+noncomputable
+def mulEquiv_of_orderOf_eq : G ≃* G' := by
+  refine MonoidHom.toMulEquiv (monoidHom_of_forall_mem_zpowers hg h.symm.dvd)
+    (monoidHom_of_forall_mem_zpowers hg' h.dvd) ?_ ?_ <;>
+  refine (MonoidHom.eq_iff_eq_on_generator (by assumption) _ _).mpr ?_ <;>
+  simp only [MonoidHom.coe_comp, Function.comp_apply, monoidHom_of_forall_mem_zpowers_apply_gen,
+    MonoidHom.id_apply]
+
+@[to_additive]
+lemma mulEquiv_of_orderOf_eq_apply_gen : mulEquiv_of_orderOf_eq hg hg' h g = g' :=
+  monoidHom_of_forall_mem_zpowers_apply_gen hg h.symm.dvd
+
+@[to_additive]
+lemma mulEquiv_of_orderOf_eq_symm_apply_gen : (mulEquiv_of_orderOf_eq hg hg' h).symm g' = g :=
+  monoidHom_of_forall_mem_zpowers_apply_gen hg' h.dvd
+
+@[to_additive]
+lemma mulEquiv_of_orderOf_eq_symm :
```

Comment:
should these be simp lemmas?

Reply:
I can make them (all four you flagged) `simp` lemmas, but I'm doubtful that `simp` will fire often due to the generating conditions.

Reply:
They are present in the LHS so I believe simp can find it.

Reply:
I take that back -- the conditions are explicit in the terms...